runtime/meta/substitution.cpp

```diff
@@ -167,52 +172,81 @@ block *substituteInternal(block *currBlock) {
   const uint64_t hdr = currBlock->h.hdr;
   uint16_t layoutInt = layout_hdr(hdr);
   if (hook_KEQUAL_eq(currBlock, to_replace)) {
+    idx2 = 0;
     if (layoutInt) {
       uint32_t tag = tag_hdr(hdr);
       uint32_t injTag = getInjectionForSortOfTag(tag);
       if (tag_hdr(replacementInj->h.hdr) != injTag) {
-        return replacementInj;
+        return incrementDebruijn(replacementInj);
       }
     }
-    return replacement;
+    return incrementDebruijn(replacement);
   }
   if (layoutInt) {
     layout *layoutData = getLayoutData(layoutInt);
     bool dirty = false;
     block *newBlock = currBlock;
+    uint32_t tag = tag_hdr(hdr);
+    std::vector<void *> arguments;
+    bool isBinder = isSymbolABinder(tag);
+    if(isBinder) {
+      idx++;
+    }
     for (unsigned i = 0; i < layoutData->nargs; i++) {
       layoutitem *argData = layoutData->args + i;
       void *arg = ((char *)currBlock) + argData->offset;
       switch(argData->cat) {
       case MAP_LAYOUT: {
         map newArg = map_map(arg, substituteInternal);
         makeDirty(dirty, argData->offset, newArg, newBlock);
-	break;
+        arguments.push_back(((char *)newBlock) + argData->offset);
+        break;
       } case LIST_LAYOUT: {
         list newArg = list_map(arg, substituteInternal);
         makeDirty(dirty, argData->offset, newArg, newBlock);
-	break;
+        arguments.push_back(((char *)newBlock) + argData->offset);
+        break;
       } case SET_LAYOUT: {
         set newArg = set_map(arg, substituteInternal);
         makeDirty(dirty, argData->offset, newArg, newBlock);
-	break;
+        arguments.push_back(((char *)newBlock) + argData->offset);
+        break;
       } case VARIABLE_LAYOUT:
         case SYMBOL_LAYOUT: {
         block *oldArg = *(block **)arg;
-	block *newArg = substituteInternal(oldArg);
+        block *newArg = substituteInternal(oldArg);
         if (oldArg != newArg || dirty) {
           makeDirty(dirty, argData->offset, newArg, newBlock);
         }
+        arguments.push_back(newArg);
         break;
       }
       case STRINGBUFFER_LAYOUT:
       case INT_LAYOUT:
       case FLOAT_LAYOUT:
+        arguments.push_back(*(void **)arg);
+        break;
       case BOOL_LAYOUT:
       default: //mint
+        arguments.push_back(arg);
         break;
       }
     }
+    if (isBinder) {
+      idx--;
+    }
+    if (isSymbolAFunction(tag)) {
+      uint64_t idx_stack = idx;
+      block *to_replace_stack = to_replace;
+      block *replacement_stack = replacement;
+      block *replacementInj_stack = replacementInj;
+      block *result = (block *)evaluateFunctionSymbol(tag, &arguments[0]);
+      to_replace = to_replace_stack;
+      replacement = replacement_stack;
+      replacementInj = replacementInj_stack;
+      idx = idx_stack;
+      return result;
+    }
```

Comment:
Are we calling this here because after going through applying substitution to all sub-terms we could now have the result be evaluatable?

I guess I'm wondering how this was done before. Is the function evaluator being called after substitution anyway? If so, is this redundant?

Reply:
no, function evaluation is not called after substitution. previously we forgot to evaluate anywhere rules after applying substitution.

Reply:
Sounds good then.